### PR TITLE
GEOMESA-2342 Add XZ2 Support for the FileSystem DataStore

### PR DESCRIPTION
**Pull request by jnh5y**
Signed-off-by: Jim Hughes <jnh5y@ccri.com>

### CODE REVIEW

File: geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-common/src/main/scala/org/locationtech/geomesa/fs/storage/common/partitions/Z2Scheme.scala [hunk 1]
```diff
@@ -8,33 +8,64 @@
...
+class XZ2Scheme(bits: Int, geom: String, leaf: Boolean) extends SpatialScheme(bits, geom, leaf) {
+  // The max XZ2 value is (4^((bits/2)+1)-1)/3.  This calculates the number of digits in that value.
+  override protected val digits: Int = math.ceil(((bits/2)+1)* math.log10(4) - math.log10(3)).toInt
+
+  override def getName: String = XZ2Scheme.Name
+
+  private val xz2 = XZ2SFC((bits / 2).asInstanceOf[Short])
+
+  override def getPartition(feature: SimpleFeature): String = {
+    // TODO support non-point geoms
```
Comment (elahrvivaz) [inline geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-common/src/main/scala/org/locationtech/geomesa/fs/storage/common/partitions/Z2Scheme.scala]: can remove this TODO I assume?

File: geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-common/src/main/scala/org/locationtech/geomesa/fs/storage/common/partitions/Z2Scheme.scala [hunk 2]
```diff
@@ -8,33 +8,64 @@
...
+  override def generateRanges(xy: Seq[(Double, Double, Double, Double)]): Seq[Long] =
+    xz2.ranges(xy).flatMap(ir => ir.lower to ir.upper)
+}
+
+abstract class SpatialScheme(bits: Int, geom: String, leaf: Boolean) extends PartitionScheme {
+
+  require(bits % 2 == 0, "Resolution must be an even number")
+
+  // TODO: The number of digits might be off for Z2 versus XZ2
+  protected val digits: Int // = math.ceil(math.log10(math.pow(2, bits))).toInt
```
Comment (elahrvivaz) [inline geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-common/src/main/scala/org/locationtech/geomesa/fs/storage/common/partitions/Z2Scheme.scala]: I assume this comment is no longer relevant and the commented out line can be removed?

File: geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-common/src/main/scala/org/locationtech/geomesa/fs/storage/common/partitions/Z2Scheme.scala [hunk 3]
```diff
@@ -50,28 +81,25 @@ class Z2Scheme(bits: Int, geom: String, leaf: Boolean) extends PartitionScheme {
     if (geometries.disjoint) {
       Collections.emptyList()
     } else {
-      val xy = geometries.values.map(GeometryUtils.bounds)
-      val enumerations = z2.ranges(xy).flatMap(ir => ir.lower to ir.upper)
+      val xy: Seq[(Double, Double, Double, Double)] = geometries.values.map(GeometryUtils.bounds)
+      val enumerations: Seq[Long] = generateRanges(xy) //z2.ranges(xy).flatMap(ir => ir.lower to ir.upper)
```
Comment (elahrvivaz) [inline geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-common/src/main/scala/org/locationtech/geomesa/fs/storage/common/partitions/Z2Scheme.scala]: can removed commented out line

**Comment by jnh5y** [comment]
Nuked the comments.

**Pull request merged**